pkg/capacity/controller.go

```diff
@@ -0,0 +1,506 @@
+	}
+)
+
+// CSICapacityClient is the relevant subset of csi.ControllerClient.
+type CSICapacityClient interface {
+	GetCapacity(ctx context.Context, in *csi.GetCapacityRequest, opts ...grpc.CallOption) (*csi.GetCapacityResponse, error)
+}
+
+// NewController creates a new controller for CSIStorageCapacity objects.
+func NewCentralCapacityController(
+	csiController CSICapacityClient,
+	driverName string,
+	client kubernetes.Interface,
+	queue workqueue.RateLimitingInterface,
+	owner metav1.OwnerReference,
+	topologyInformer topology.Informer,
+	scInformer storageinformersv1.StorageClassInformer,
+	enumerateObjects int,
+) *Controller {
+	c := &Controller{
+		csiController:    csiController,
+		driverName:       driverName,
+		client:           client,
+		queue:            queue,
+		owner:            owner,
+		topologyInformer: topologyInformer,
+		scInformer:       scInformer,
+		capacities:       map[workItem]*storagev1alpha1.CSIStorageCapacity{},
+		enumerateObjects: enumerateObjects,
+	}
+
+	// Now register for changes. Depending on the implementation of the informers,
+	// this may already invoke callbacks.
+	handler := cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.onSCAddOrUpdate(obj.(*storagev1.StorageClass)) },
+		UpdateFunc: func(_ interface{}, newObj interface{}) { c.onSCAddOrUpdate(newObj.(*storagev1.StorageClass)) },
+		DeleteFunc: func(obj interface{}) { c.onSCDelete(obj.(*storagev1.StorageClass)) },
+	}
+	c.scInformer.Informer().AddEventHandler(handler)
+	c.topologyInformer.AddCallback(c.onTopologyChanges)
+
+	return c
+}
+
+// Run is a main Controller handler
+func (c *Controller) Run(ctx context.Context, threadiness int) {
+	klog.Info("Starting Capacity Controller")
+	defer utilruntime.HandleCrash() // TODO: do we need this?
+	defer c.queue.ShutDown()
+	go c.scInformer.Informer().Run(ctx.Done())
+	go c.topologyInformer.Run(ctx)
+
+	c.prepare(ctx)
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(func() {
+			c.runWorker(ctx)
+		}, time.Second, ctx.Done())
+	}
+
+	klog.Info("Started Capacity Controller")
+	<-ctx.Done()
+	klog.Info("Shutting down Capacity Controller")
+}
+
+func (c *Controller) prepare(ctx context.Context) {
+	// Wait for topology and storage class informer sync. Once we have that,
+	// we know which CSIStorageCapacity objects we need.
+	if !cache.WaitForCacheSync(ctx.Done(), c.topologyInformer.HasSynced, c.scInformer.Informer().HasSynced) {
+		return
+	}
+	klog.V(3).Infof("Initial number of topology segments %d, storage classes %d, potential CSIStorageCapacity objects %d",
+		len(c.topologyInformer.List()),
+		func() int {
+			scs, _ := c.scInformer.Lister().List(labels.Everything())
+			return len(scs)
+		}(),
+		len(c.capacities))
+
+	// Retrieve all existing CSIStorageCapacity objects that are owned by us
+	// *before* working on items (because that would then create duplicates).
+	// There's no reconciliation at runtime because changes not made by us
+	// should never happen.
```

Comment:
I'm undecided whether I should use a CSIStorageCapacity informer or just list them periodically.

If I use an informer, the controller will get all of its own changes sent back to it, which adds a permanent, fixed overhead. The advantage is that the controller can react to undesirable changes quickly.

If I list, probably with a long delay, then the overhead might be lower (depending on how often it polls), but undesirable changes will be detected more slowly.

@jsafrane : What would you prefer?

Reply:
@jsafrane: ping? ^^^

Reply:
I've convinced myself that an informer is better because:
- a constant overhead is simpler to measure and plan for
- the code for handling the update conflict can be removed

Implemented (but not pushed yet).